stan/math/prim/mat/prob/categorical_logit_glm_lpmf.hpp

```diff
@@ -0,0 +1,206 @@
+  const auto& beta_val = value_of_rec(beta);
+  const auto& alpha_val = value_of_rec(alpha);
+
+  const auto& alpha_val_vec = as_column_vector_or_scalar(alpha_val).transpose();
+
+  Array<T_partials_return, T_x_rows, Dynamic> lin
+      = (x_val * beta_val).rowwise() + alpha_val_vec;
+  Array<T_partials_return, T_x_rows, 1> lin_max
+      = lin.rowwise()
+            .maxCoeff();  // This is used to prevent overflow when
+                          // calculating softmax/log_sum_exp and similar
+                          // expressions
+  Array<T_partials_return, T_x_rows, Dynamic> exp_lin
+      = exp(lin.colwise() - lin_max);
+  Array<T_partials_return, T_x_rows, 1> inv_sum_exp_lin
+      = 1 / exp_lin.rowwise().sum();
+
+  T_partials_return logp = log(inv_sum_exp_lin).sum() - lin_max.sum();
+  if (T_x_rows == 1) {
+    logp *= N_instances;
+  }
+  scalar_seq_view<T_y> y_seq(y);
+  for (int i = 0; i < N_instances; i++) {
+    if (T_x_rows == 1) {
+      logp += lin(0, y_seq[i] - 1);
+    } else {
+      logp += lin(i, y_seq[i] - 1);
+    }
+  }
+  // TODO(Tadej) maybe we can replace previous block with the following line
+  // when we have newer Eigen  T_partials_return logp =
+  // lin(Eigen::all,y-1).sum() + log(inv_sum_exp_lin).sum() - lin_max.sum();
+
+  if (!std::isfinite(logp)) {
+    check_finite(function, "Weight vector", beta);
+    check_finite(function, "Intercept", alpha);
+    check_finite(function, "Matrix of independent variables", x);
+  }
+
+  // Compute the derivatives.
+  operands_and_partials<Matrix<T_x_scalar, T_x_rows, Dynamic>,
+                        Matrix<T_alpha_scalar, Dynamic, 1>,
+                        Matrix<T_beta_scalar, Dynamic, Dynamic>>
+      ops_partials(x, alpha, beta);
+
+  if (!is_constant_struct<T_x_scalar>::value) {
+    if (T_x_rows == 1) {
+      Array<double, 1, Dynamic> beta_y = beta_val.col(y_seq[0] - 1);
+      for (int i = 1; i < N_instances; i++) {
+        beta_y += beta_val.col(y_seq[i] - 1).array();
+      }
+      ops_partials.edge1_.partials_
+          = beta_y
+            - (exp_lin.matrix() * beta_val.transpose()).array().colwise()
+                  * inv_sum_exp_lin * N_instances;
+    } else {
+      Array<double, Dynamic, Dynamic> beta_y(N_instances, N_attributes);
+      for (int i = 0; i < N_instances; i++) {
+        beta_y.row(i) = beta_val.col(y_seq[i] - 1);
+      }
+      ops_partials.edge1_.partials_
+          = beta_y
+            - (exp_lin.matrix() * beta_val.transpose()).array().colwise()
+                  * inv_sum_exp_lin;
+      // TODO(Tadej) maybe we can replace previous block with the following line
+      // when we have newer Eigen  ops_partials.edge1_.partials_ = beta_val(y -
+      // 1, all) - (exp_lin.matrix() * beta.transpose()).colwise() *
+      // inv_sum_exp_lin;
+    }
+  }
+  if (!is_constant_struct<T_alpha_scalar>::value
+      || !is_constant_struct<T_beta_scalar>::value) {
+    Array<T_partials_return, T_x_rows, Dynamic> neg_softmax_lin
```

Comment:
For Array (which I believe is from Eigen), do we want to include prim/Eigen.hpp?

Any benefits of using an Array over matrix? (Array I guess has more dimensions than 2...)

Reply:
`Array` and `Matrix` types both have 2 dimensions. They only differ in what multiplication does to them. For `Matrix` it is matrix multiplication and for `Array` it is element-wise multiplication. 

Which header do you mean? There is no Eigen.hpp in stan/math/prim. `Array` comes from Eigen/Core.

Reply:
In prim/mat.hpp there's an Eigen.hpp, which includes core.

Not at all necessary, as you'd including stuff you're not using anyway.